byway/src/main/webapp/base.js

```diff
@@ -92,3 +93,51 @@ function findPlace(placeId, placesService) {
   return result;
 }
 
+/**
+ * Class used to throw error in getDirections() with both message and status
+ */
+class MapStatusError extends Error {
+    constructor(message, status) {
+        super(message);
+        this.name = 'MapStatusError';
+        this.status = status;
+    }
+}
+
+/**
+ * Makes a request to the Directions API to return information about the route
+ * @param {DirectionsService} directionsService object that communicates with the GMaps API service
+ * @param {directionsRenderer} directionsRenderer object that renders display results on the map
+ * @param {String} start placeId as string
+ * @param {String} end placeId as string
+ * @param {Array} [waypoints] array of waypoint objects
+ * @return {Promise} result Directions route object with fields route, legs, etc.
+ */
+function getRouteForTrip(directionsService, directionsRenderer, start, end, waypoints) {
+   let request = {
+    origin:  {placeId : start},
+    destination: {placeId : end},
+    waypoints, 
+    travelMode: 'DRIVING',
+    optimizeWaypoints: true
+  };
+  const result = new Promise((resolve,reject) => {
+    directionsService.route(request, (result, status) => {
+      if (status == 'OK') {
+        directionsRenderer.setDirections(result);
```

Comment:
Try to keep the operation of converting `directionService.route` into a promise from what you want to do with the results. 

I'd split this in 3 functions:

1. Wraps `directionsService.route` and makes it return a promise. something like `computeRoute(request, directionsService)`
2. Handles getting the route for the current trip by calling that function. like `computeRouteForTrip(directionsService, start, end, waypoints)`, which just takes care of building the request and calling `computeRoute`.
3. Chaining off of or awaiting `computeRouteForTrip` to call `directionsRenderer.setDirections`

Reply:
Thanks! I've separated getRouteForTrip() into two separate functions for cleaner structure and reuse.